src/net/sourceforge/kolmafia/textui/langserver/textdocumentservice/AshTextDocumentService.java

```diff
@@ -30,14 +37,55 @@ public final void setCapabilities(final ServerCapabilities capabilities) {
   }
 
   @Override
-  public void didOpen(DidOpenTextDocumentParams params) {}
+  public void didOpen(DidOpenTextDocumentParams params) {
+    this.parent.executor.execute(
+        () -> {
+          TextDocumentItem document = params.getTextDocument();
+
+          File file = FilesMonitor.UriToFile(document.getUri());
+
+          this.parent.monitor.updateFile(file, document.getText(), document.getVersion());
+        });
+  }
 
   @Override
-  public void didChange(DidChangeTextDocumentParams params) {}
+  public void didChange(DidChangeTextDocumentParams params) {
+    this.parent.executor.execute(
+        () -> {
+          VersionedTextDocumentIdentifier document = params.getTextDocument();
+          List<TextDocumentContentChangeEvent> changes = params.getContentChanges();
+
+          if (changes.size() == 0) {
+            // Nothing to see here
+            return;
+          }
+
+          File file = FilesMonitor.UriToFile(document.getUri());
+
+          // We don't support incremental changes, so we expect the client
+          // to put the whole file's content in a single TextDocumentContentChangeEvent
+          this.parent.monitor.updateFile(file, changes.get(0).getText(), document.getVersion());
+        });
+  }
 
   @Override
-  public void didClose(DidCloseTextDocumentParams params) {}
+  public void didClose(DidCloseTextDocumentParams params) {
+    this.parent.executor.execute(
+        () -> {
+          TextDocumentIdentifier document = params.getTextDocument();
+
+          File file = FilesMonitor.UriToFile(document.getUri());
+
+          this.parent.monitor.updateFile(file, null, -1);
+        });
+  }
 
   @Override
-  public void didSave(DidSaveTextDocumentParams params) {}
+  public void didSave(DidSaveTextDocumentParams params) {
+    // What we had in memory should now match the content of the file.
+    // Should we just ignore these notifications?
+    // We could set the Script's text and version back to null and -1, respectively. If so, either
+    // A- the two are indeed identical, and this is just a waste of processing time, or
+    // B- the two are different, which should not happen; the issue to solve would be elsewhere...
```

Comment:
This sounds like a design question which should be sorted as part of this PR.

I think ignoring the notifications is fine.